conmon-rs/server/src/oom_watcher.rs

```diff
@@ -0,0 +1,314 @@
+use anyhow::{anyhow, Context, Result};
+use lazy_static::lazy_static;
+use log::{debug, error};
+use nix::sys::statfs::{statfs, FsType};
+use notify::{Error, Event, RecommendedWatcher, RecursiveMode, Watcher};
+use regex::Regex;
+use std::os::unix::prelude::AsRawFd;
+use std::path::{Path, PathBuf};
+use tokio::fs::File;
+use tokio::io::{AsyncBufReadExt, AsyncReadExt, AsyncWriteExt, BufReader};
+use tokio::sync::mpsc::{channel, Receiver};
+use tokio::task::{self, JoinHandle};
+use tokio_eventfd::EventFd;
+use tokio_util::sync::CancellationToken;
+
+#[cfg(any(all(target_os = "linux", target_env = "musl")))]
+pub const CGROUP2_SUPER_MAGIC: FsType = FsType(libc::CGROUP2_SUPER_MAGIC as u64);
+#[cfg(any(all(target_os = "linux", not(target_env = "musl"))))]
+pub const CGROUP2_SUPER_MAGIC: FsType = FsType(libc::CGROUP2_SUPER_MAGIC as i64);
+
+static CGROUP_ROOT: &str = "/sys/fs/cgroup";
+
+lazy_static! {
+    static ref IS_CGROUP_V2: bool = {
+        if let Ok(sts) = statfs(CGROUP_ROOT) {
+            return sts.filesystem_type() == CGROUP2_SUPER_MAGIC;
+        }
+        false
+    };
+}
+
+pub struct OOMWatcher {
+    pid: u32,
+    token: CancellationToken,
+    task: JoinHandle<()>,
+}
+
+pub struct OOMEvent {
+    pub oom: bool,
+}
+
+impl OOMWatcher {
+    pub async fn new(
+        pid: u32,
+        exit_paths: &[PathBuf],
+        tx: tokio::sync::mpsc::Sender<OOMEvent>,
+    ) -> OOMWatcher {
+        let token = CancellationToken::new();
+        let exit_paths = exit_paths.to_owned();
+        let task = {
+            let stop = token.clone();
+            task::spawn(async move {
+                if *IS_CGROUP_V2 {
+                    let _ = Self::oom_handling_cgroup_v2(stop, pid, &exit_paths, tx)
+                        .await
+                        .context("setup cgroupv2 oom handling");
+                } else {
+                    let _ = Self::oom_handling_cgroup_v1(stop, pid, &exit_paths, tx)
+                        .await
+                        .context("setup cgroupv1 oom handling");
+                }
```

Comment:
I'm wondering why we hide the error here:
```suggestion
                if let Err(e) = if *IS_CGROUP_V2 {
                    Self::oom_handling_cgroup_v2(stop, pid, &exit_paths, tx)
                        .await
                        .context("setup cgroupv2 oom handling")
                } else {
                    Self::oom_handling_cgroup_v1(stop, pid, &exit_paths, tx)
                        .await
                        .context("setup cgroupv1 oom handling")
                } {
                    error!("Failed to watch OOM: {:#}", e)
                }

```

Reply:
fixed.